### PR TITLE
UserInfoFetcher: ActiveDirectory backend

### LINKED ISSUE #524: Support ActiveDirectory backend for user-info-fetcher
**Background**
We currently only support KeyClouk as a backend for the UserInfoFetcher. This allows pull through access to Active Directory in theory, but is an extra component to manage, that can be unneeded, when _everything_ should come from AD anyway.

We should look into supporting Active Directory as a direct backend for the UserInfofetcher.

**Implementation Tasks**
- [ ] https://github.com/stackabletech/opa-operator/pull/622

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Fixes #524 

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

**Comment by fhennig** [comment]
I saw this in the decision RfC column, anything particular you are looking for comments on?

**Comment by nightkr** [comment]
Primarily the CRD changes adding the new backend.

**Comment by nightkr** [comment]
Moving CRD review to _voting_, please vote on this comment.

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc [hunk 1]
```diff
@@ -46,21 +46,54 @@ stringData:
...
 
 NOTE: The OAuth2 Client in Keycloak must be given the `view-users` _Service Account Role_ for the realm that the users are in.
 
+[#backend-activedirectory]
+== Active Directory
+
+WARNING: The Active Directory backend is experimental, and subject to change.
+
+Fetches user attributes and groups over LDAP.
+
```
Comment (sbernauer) [inline docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:72]: Under the assumption this is correct
```suggestion

For this to work user-info-fetcher needs to be provided with a Kerberos keytab that enables it to access Active Directory.
This is provided by a configurable SecretClass.
```

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc [hunk 2]
```diff
@@ -46,21 +46,54 @@ stringData:
...
 <3> Obtain Keycloak API credentials from the specified secret. The Secret must have `clientId` and `clientSecret` entries.
 <4> Refer to the applicable realm in your Keycloak server.
 
 Currently the following backends are supported:
 
-* <<_keycloak>>
+* xref:#backend-keycloak[]
+* xref:#backend-activedirectory[]
 
+[#backend-keycloak]
```
Comment (sbernauer) [inline docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:58]: I know it was this way before, but IMHO it makes sense to move the Keycloak example above below the Keycloak section :)

File: rust/crd/src/user_info_fetcher.rs [hunk 3]
```diff
@@ -88,6 +94,27 @@ fn aas_default_port() -> u16 {
     5000
 }
 
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ActiveDirectoryBackend {
+    /// Hostname of the domain controller, e.g. `ad-ds-1.contoso.com`.
+    pub ldap_hostname: String,
```
Comment (sbernauer) [inline rust/crd/src/user_info_fetcher.rs:101]: In the AuthenticationClass we also have an `ldapPort` field: https://docs.stackable.tech/home/stable/concepts/authentication#ldap. It might make sense to have that for consistency reasons as well.

If not please test a non-default port and update the CRD docs, e.g.
```
Hostname of the domain controller, e.g. `ad-ds-1.contoso.com` or `ad-ds-1.contoso.com:1234`.
```

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc (hunk 2 above)
Comment (nightkr) [reply docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:63]: The API docs are fairly independent from the backend in use, mentioned that the example shows some keycloakisms in db442967e6b29d320c8fd66997b615c7ab8551dc

File: rust/crd/src/user_info_fetcher.rs (hunk 3 above)
Comment (nightkr) [reply rust/crd/src/user_info_fetcher.rs:101]: That's fair, the inspiration here was secret-op's LDAP support, which also doesn't support custom LDAP ports at the moment. As far as I know, there's no way to customize AD's LDAP port.

Comment (nightkr) [reply rust/crd/src/user_info_fetcher.rs:101]: In fairness, there are a few things we could port over from secret-op here:

- rename to `ldapServer`
- explicitly forbid port numbers (though it should break the SASL bind anyway)

Comment (sbernauer) [reply rust/crd/src/user_info_fetcher.rs:101]: > As far as I know, there's no way to customize AD's LDAP port.

Oh that's a bit unexpected :D Happy to rename it to `ldapServer` and close this discussion

Comment (nightkr) [reply rust/crd/src/user_info_fetcher.rs]: 94febb5bcec447bdcd80041585743e20b7cb4883

**Comment by nightkr** [comment]
Closing CRD vote as _accepted_

**Comment by nightkr** [comment]
Proper testing is blocked on https://github.com/stackabletech/issues/issues/629

File: rust/user-info-fetcher/src/backend/active_directory.rs [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+const LDAP_MATCHING_RULE_IN_CHAIN: &str = ":1.2.840.113556.1.4.1941:";
+
+const LDAP_FIELD_OBJECT_ID: &str = "objectGUID";
+const LDAP_FIELD_OBJECT_SECURITY_ID: &str = "objectSid";
+const LDAP_FIELD_OBJECT_DISTINGUISHED_NAME: &str = "dn";
+const LDAP_FIELD_USER_NAME: &str = "userPrincipalName";
+const LDAP_FIELD_USER_PRIMARY_GROUP_RID: &str = "primaryGroupID";
+const LDAP_FIELD_GROUP_MEMBER: &str = "member";
+
+pub(crate) async fn get_user_info(
```
Comment (soenkeliebau) [inline rust/user-info-fetcher/src/backend/active_directory.rs:93]: Is this not instrumented for tracing on purpose?

File: rust/user-info-fetcher/src/backend/active_directory.rs [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    let requested_user_attrs = [
+        LDAP_FIELD_OBJECT_SECURITY_ID,
+        LDAP_FIELD_OBJECT_ID,
+        LDAP_FIELD_USER_NAME,
+        LDAP_FIELD_USER_PRIMARY_GROUP_RID,
+    ]
+    .into_iter()
+    .chain(custom_attribute_mappings.values().map(String::as_str))
+    .collect::<Vec<&str>>();
+    let user = ldap
```
Comment (soenkeliebau) [inline rust/user-info-fetcher/src/backend/active_directory.rs:142]: Would it make sense to add a debug/trace statement here with what attributes where requested and what was returned? 
I manually added these while reviewing and debugging why it didn't work.

File: rust/user-info-fetcher/src/backend/active_directory.rs [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+        id: id.map(|id| id.to_string()),
+        username,
+        groups,
+        custom_attributes,
+    })
+}
+
+/// Gets the distinguished names of all of `user`'s groups, both primary and secondary.
+#[tracing::instrument(skip(ldap, base_dn, user, user_sid))]
+async fn user_group_distinguished_names(
```
Comment (soenkeliebau) [inline rust/user-info-fetcher/src/backend/active_directory.rs:243]: Would it make sense to add tests for this at some point? Doesn't need to be before the release, I dragged my feet reviewing this long enough to make that non-viable any more, but maybe in a later PR?

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc [hunk 7]
```diff
@@ -46,21 +46,61 @@ stringData:
...
 
+[#backend-activedirectory]
+=== Active Directory
+
+WARNING: The Active Directory backend is experimental, and subject to change.
+
+Fetches user attributes and groups over LDAP.
+
+For this to work user-info-fetcher needs to be provided with a Kerberos keytab that enables it to access Active Directory.
+This is provided by a configurable SecretClass.
```
Comment (soenkeliebau) [inline docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:79]: custom attributes are completely missing from the docs I think? Is that on purpose to discourage using them, or can we maybe add them to the example at least?

File: rust/user-info-fetcher/src/utils/tls.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+// NOTE: MUST be kept in sync with all configure_* functions
+pub async fn configure_native_tls(
+    tls: &TlsClientDetails,
+) -> Result<native_tls::TlsConnector, Error> {
+    let mut builder = native_tls::TlsConnector::builder();
+    if tls.uses_tls() && !tls.uses_tls_verification() {
+        builder.danger_accept_invalid_certs(true);
+    } else if let Some(tls_ca_cert_mount_path) = tls.tls_ca_cert_mount_path() {
+        builder.disable_built_in_roots(true);
+        // native-tls doesn't support parsing CA *bundles*, so split them using rustls first
```
Comment (soenkeliebau) [inline rust/user-info-fetcher/src/utils/tls.rs:60]: Does this _only_ affect us here, or is it something that we just hit here first? aka - worth pulling something like that into op-rs?

**Review by soenkeliebau — APPROVED** [review]
Looked over the code and tested it - works as intended!

So, lgtm in principle, I just left a few questions throughout the code, none of them showstoppers though.

File: rust/user-info-fetcher/src/backend/active_directory.rs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+            Error::ParseUserSid { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+// Matching rules defined at https://learn.microsoft.com/en-us/windows/win32/adsi/search-filter-syntax#operators
+/// Makes DN filters apply recursively to group membership
+const LDAP_MATCHING_RULE_IN_CHAIN: &str = ":1.2.840.113556.1.4.1941:";
+
+const LDAP_FIELD_OBJECT_ID: &str = "objectGUID";
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/user-info-fetcher/src/backend/active_directory.rs:86]: AFAICT, the special ones seem to be fairly hard-coded (https://learn.microsoft.com/en-us/windows/win32/ad/naming-properties for example), what I do remember seeing is different orgs disagreeing on how to store data that's outside of the native AD data model (like a swedish personnummer).

If the need shows up we can always go back and make them more configurable when we understand the area better.

File: rust/user-info-fetcher/src/backend/active_directory.rs (hunk 4 above)
Comment (nightkr) [reply rust/user-info-fetcher/src/backend/active_directory.rs:93]: No good reason, that makes sense to add.

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc (hunk 7 above)
Comment (nightkr) [reply docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:79]: Mostly a case of when what was written I think.

File: rust/user-info-fetcher/src/utils/tls.rs (hunk 8 above)
Comment (nightkr) [reply rust/user-info-fetcher/src/utils/tls.rs:60]: Probably worth upstreaming at some point, yeah.

File: rust/user-info-fetcher/src/backend/active_directory.rs (hunk 4 above)
Comment (nightkr) [reply rust/user-info-fetcher/src/backend/active_directory.rs:94]: 3fc772a478228e120a65ef5058beda8385151eaa

File: rust/user-info-fetcher/src/backend/active_directory.rs (hunk 5 above)
Comment (nightkr) [reply rust/user-info-fetcher/src/backend/active_directory.rs:149]: 3fc772a478228e120a65ef5058beda8385151eaa

File: rust/user-info-fetcher/src/backend/active_directory.rs (hunk 6 above)
Comment (nightkr) [reply rust/user-info-fetcher/src/backend/active_directory.rs:261]: We need to think about nice ways to set it up, but ideally yes.

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc (hunk 7 above)
Comment (nightkr) [reply docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:79]: 77b8bc1f2ec6a0ff27f24d5c175500a4033e8669

File: rust/user-info-fetcher/src/backend/active_directory.rs (hunk 6 above)
Comment (soenkeliebau) [reply rust/user-info-fetcher/src/backend/active_directory.rs:261]: I was kinda hoping that we'd be able to find a list of a few hundred known special and edge cases with proper parsing that we can just run this against as a unit test ..
I'll see myself out :)

Comment (soenkeliebau) [reply rust/user-info-fetcher/src/backend/active_directory.rs:261]: https://github.com/stackabletech/opa-operator/issues/640

File: rust/user-info-fetcher/src/utils/tls.rs (hunk 8 above)
Comment (soenkeliebau) [reply rust/user-info-fetcher/src/utils/tls.rs:60]: Fine for me .. random thoughts, are we tracking candidates for upstreaming somewhere? I think you just created a board for deprecated stuff, a similar mechanism might be useful?

**Review by soenkeliebau — APPROVED** [review]

**Review by soenkeliebau — APPROVED** [review]

File: docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc (hunk 2 above)
Comment (sbernauer) [reply docs/modules/opa/pages/usage-guide/user-info-fetcher.adoc:64]: Thanks
